Refactor/use iterators to preselect utxos

nymius opened the pull request:
### Description

There were multiple calls for de-duplication of selected UTxOs in `Wallet::create_tx`: ([1](https://github.com/bitcoindevkit/bdk/blob/abc305612160c0e2ce85c7bba4c2c162ff488adc/crates/wallet/src/wallet/mod.rs#L2016-L2020)) and ([2](https://github.com/bitcoindevkit/bdk/blob/abc305612160c0e2ce85c7bba4c2c162ff488adc/crates/wallet/src/wallet/mod.rs#L1452)).

As the test [`test_filter_duplicates`](https://github.com/bitcoindevkit/bdk/blob/master/crates/wallet/src/wallet/coin_selection.rs#L1666-L1695) shows, there are four possible cases for duplication of UTxOs while feeding the coin selection algorithms.

1. no duplication: out of concern
2. duplication in the required utxos only: covered by the source of `required_utxos`, `Wallet::list_unspent`, which [roots back the provided `UTxOs` to a `HashMap`](https://github.com/bitcoindevkit/bdk/blob/a5335a18431dfecea1f524af44c953b79a96867c/crates/chain/src/tx_graph.rs#L911-L912) which should avoid any duplication by definition
3. duplication in the optional utxos only: is the only one possible as optional `UTxOs` are stored in a `Vec` and no checks are performed about the duplicity of their members.
4. duplication across the required and optional utxos: is already covered by `Wallet::preselect_utxos`, which avoid the processing of required UTxOs while listing the unspent available UTxOs in the wallet.

This refactor does the following:
- Refactors `TxParams::utxos` type to be `HashSet<LocalOutput>` avoiding the duplication case 3
- Keeps avoiding case 4 without overhead and allowing a query closer to O(1) on avg. to cover duplication case 4 (before was O(n) where n is the size of required utxos) thanks to the above change.
- Still covers case 2 because the [source of `required_utxos`, `Wallet::list_unspent`](https://github.com/bitcoindevkit/bdk/blob/a5335a18431dfecea1f524af44c953b79a96867c/crates/chain/src/tx_graph.rs#L911-L912) comes from a `HashMap` which should avoid duplication by definition.
- Moves the computation of the `WeightedUtxos` to the last part of UTxO filtering, allowing the unification of the computation for local outputs.
- Removes some extra allocations done for helpers structures or intermediate results while filtering UTxOs.
- Allows for future integration of UTxO filtering methods for other utilities, e.g.: filter locked utxos
  ```rust
  .filter(|utxo| !self.is_utxo_locked(utxo.outpoint, self.latest_checkpoint().height()))
  ```
- Adds more comments for each filtering step.
- Differentiates `foreign_utxos`, which should include a provided satisfation weight to use them effectively, and `utxos`, manually selected UTxOs for which the wallet can compute their satisfaction weight without external resources.

With these changes all four cases would be covered, and `coin_selection::filter_duplicates` is no longer needed.

Fixes bitcoindevkit/bdk#1794.

### Notes to the reviewers

I added three test to cover the interesting cases for duplication:
    - there are no duplicates in required utxos
    - there are no duplicates in optional utxos
    - there are no duplicates across optional and required utxos
the three of them have been prefixed with `not_duplicated_utxos*` to allow its joint execution under the command:
```bash
cargo test -- not_duplicated_utxos
```
because the guarantees for the three conditions above are spread in different parts of the code.

### Changelog notice

No changes to public APIs.

### Checklists

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing
* [x] This pull request does not break the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

nymius committed:
refactor(wallet): use iterators and adaptors in preselect_utxos

There were multiple calls for de-duplication of selected UTxOs.

As the test `test_filter_duplicates` shows, there are four possible
cases for duplication of UTxOs while feeding the coin selection
algorithms.

1. no duplication: out of concern
2. duplication in the required utxos only: covered by the source of
   `required_utxos`, `Wallet::list_unspent`, which roots back the
   provided `UTxOs` to a `HashMap` which should avoid any duplication by
   definition
3. duplication in the optional utxos only: is the only one possible as
   optional `UTxOs` are stored in a `Vec` and no checks are performed
   about the duplicity of their members.
4. duplication across the required and optional utxos: is already
   covered by `Wallet::preselect_utxos`, which avoid the processing of
   required UTxOs while listing the unspent available UTxOs in the
   wallet.

This refactor changes:
- `TxParams::utxos` type to be `HashSet<LocalOutput>` avoiding the
  duplication case 3, and allowing a query closer to O(1) on avg. to
  cover duplication case 4 (before was O(n) where n is the size of
  required utxos).
- Moves the computation of the `WeightedUtxos` to the last part of UTxO
  filtering, allowing the unification of the computation for local
  outputs.
- Removes some extra allocations done for helpers structures or
  intermediate results while filtering UTxOs.
- Allows for future integration of UTxO filtering methods for other
  utilities.
- Adds more comments for each filtering step.

With these changes all four cases would be covered, and
`coin_selection::filter_duplicates` would be no longer needed.

nymius committed:
refactor(wallet): remove coin_selection::filter_duplicates

evanlinjin commented:
Hey @nymius, does this actually fix something or is this purely a refactor?

Note that we want to redo the tx building / creating logic to use `bdk_coin_select` and miniscript planning module so I think it's good policy to only do changes that are fixes or implement features that users are requesting.

evanlinjin commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1976,116 +1968,85 @@ impl Wallet {
+            // NOTE: this avoid UTxOs in both required and optional list
+            .filter(|may_spend| !params.utxos.contains(may_spend))
+            // only process optional UTxOs if manually_selected_only is false
+            .take_while(|_| !params.manually_selected_only)
```
I think you are trying to be too fancy here. It'll be easier to read if we use an if statement and have `may_spend` as a separate variable.

```rust

let may_spend = if params.manually_selected_only {
    // This will be empty
} else {
    self.list_unspent()
    // etc...
};
```

nymius replied:
Done at e78b0e60b0b43579eb87ee3b4c746f2d3e62d65a

nymius commented:
> Hey @nymius, does this actually fix something or is this purely a refactor?

Hi @evanlinjin. Yes, maybe I went to far with the changes, still, it fixes the duplicate issue I open in bitcoindevkit/bdk#1794 by using a `HashSet` to handle manually selected UTxOs.

Then, about the refactor:

The research of the code push me to isolate the pre selection steps as items on a checklist and that's why It ended up as a refactor.

My idea is to further isolate each filter in their own iterator adaptor (a separated function for each one) that consumes `impl Iterator<Item = LocalOutputs>` and returns the same type, to later also allow the introduction of custom filters in `TxParams` which I think can be helpful to implement checks on the ability to spend a Taproot output, for example, in Silent Payments.

The heavy use of iterators came for trying to avoid the allocation of helpers, like `satisfies_confirmed`.

I envisioned something like this:
```rust
let optional_utxos = self
    .list_unspent()
    .check_are_not_already_manually_selected()
    .check_are_not_unspendable()
    .check_confirmed_only_if_RBF()
    .check_is_local_utxo()
    .check_is_mature_if_coinbase();
// then
let (required, optional) = optional_utxos.chain(required_utxos.iter().clone())
    .get_weighted_utxos()
    .chain(foreign_utxos.iter().clone())
    .apply_custom_validation_for_all_tx_inputs()
    .split_utxos_in_required_and_optional()
```

---

Discussing this today with @ValuedMammal, I decided to do the following changes:
1. [x] Just return optional utxos as they are the only needed values.
2. [x] Move the calculus of `WeightedUtxos` to another method. **Placed inside create_tx for now**
3. [x] Rename `preselect_utxos` to `filter_utxos` to correct the semantics the above point will change.
4. [x] Implement unit test for the feature to make the PR sound.
5. [x] Make `current_height` parameter not optional.

If we don't agree on the above, I propose the following alternatives:
- keep points 1 and 2 above.
- keep the old filtering implementation in place.
- keep the changes to the `build_fee_bump` (new `foreign_utxos` field in `TxParams`).
- drop all the other changes.

nymius committed:
test(wallet): check there are no duplicates across required and optional utxos

This test replaces the one used to test
`coin_selection::filter_duplicates` introduced in
5299db34cb9117ad1b66a6afcb51f6ca7e1f0d95.

As the code changed and there is not a single point to verificate the
following properties:
- there are no duplicates in required utxos
- there are no duplicates in optional utxos
- there are no duplicates across optional and required utxos
anymore, test have been prefixed with `not_duplicated_utxos*` to allow
its joint execution by using the following command:
cargo test -- not_duplicated_utxos

nymius commented:
Rebased

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1976,116 +1993,52 @@ impl Wallet {
-
-        if must_use_all_available {
-            must_spend.append(&mut may_spend);
+    fn filter_utxos<'a>(&'a self, params: &'a TxParams, current_height: u32) -> Vec<WeightedUtxo> {
```
The lifetime `'a` doesn't seem to be needed.

nymius replied:
Removed in 4c28ba52c60415f66b2c32238bc83c61d5a18df3

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2604,3 +2557,137 @@ macro_rules! doctest_wallet {
+
+        for _ in 0..3 {
+            insert_tx(&mut wallet, two_output_tx.clone());
+        }
```
Inserting the same tx again doesn't add more txs to the wallet.

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2604,3 +2557,137 @@ macro_rules! doctest_wallet {
+            .collect::<Vec<_>>();
+        // notice expected doesn't include the first output from two_output_tx as it should be
+        // filtered out
+        let expected = vec![OutPoint { txid, vout: 0 }, OutPoint { txid, vout: 1 }];
```
I'm confused by this comment, is it left over from the other test?

I think this test would be better if we add a bunch of utxos (~10) to the wallet, then call `filter_utxos` and check that 1) none are duplicated and 2) none that we expect to be present are filtered out by accident.

nymius replied:
> I'm confused by this comment, is it left over from the other test?

I'm confused too, probably I changed the test after the comment.

> 
> I think this test would be better if we add a bunch of utxos (~10) to the wallet, then call `filter_utxos` and check that 1) none are duplicated and 2) none that we expect to be present are filtered out by accident.

What I wanted to prove with the tests I implemented is that the only needed test case for `filter_utxos` is when there are duplicated utxos across optional and required.
In order to do that, here, I wanted to prove is impossible for `filter_chain_unspent` to produce duplicated optional utxos. That's why I mistakenly tried to insert the same tx multiple times. The thing is I have no knowledge of any way to incorporate duplicated utxos in the wallet bypassing those checks, and as the list is generated from an iterator on demand, the only way I can think right now is some kind of mock for the method, as it isn't an input of `filter_utxos`.

nymius replied:
Will remove the test for now, as I don't find a deterministic way to produce duplicated utxos which will be de duplicated later by wallet logic.
In relation to 2) I think [this line](https://github.com/bitcoindevkit/bdk/pull/1798/files#diff-641f4600ade27412ccd63976bcbff11451a7b34703e84d953f864d6cb02a2290R2616) should be enough to assert that.

nymius replied:
Test removed in fe3c17e513b41ab3e5933e7bc45b8e1e865931ec

ValuedMammal commented on crates/wallet/tests/wallet.rs:
```diff
@@ -3875,8 +3875,8 @@ fn test_spend_coinbase() {
-    let not_yet_mature_time = confirmation_height + COINBASE_MATURITY - 1;
-    let maturity_time = confirmation_height + COINBASE_MATURITY;
+    let not_yet_mature_time = confirmation_height + COINBASE_MATURITY - 2;
+    let maturity_time = confirmation_height + COINBASE_MATURITY - 1;
```
A code comment might help to explain why this change

nymius replied:
Commented at e8e21e1178a8b7bd781611b50d3ac5962135c241

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1618,60 +1640,54 @@ impl Wallet {
+                                    outpoint: txin.previous_output,
+                                    sequence: txin.sequence,
+                                    psbt_input: Box::new(psbt::Input {
+                                        witness_utxo: Some(txout.clone()),
```
I know this was here already but shouldn't we only set `witness_utxo` if `txout.script_pubkey.witness_version().is_some()` ?

nymius replied:
Is a good check. I assume `witness_utxos` and `non_witness_utxos` are kind of exclusive, unless we don't know the nature of the `UTxO` being spend, right?

nymius replied:
Check added in c4f7c0e273bef379bf76469db7921649288897eb

ValuedMammal commented:
📌  I don't think we've considered what would happen if a third party provides us with a "foreign" utxo that actually duplicates a utxo owned by the wallet, potentially causing us to sign something we didn't agree to. It might need more research so I'll open an issue. bitcoindevkit/bdk_wallet#29

nymius commented:
> 📌 I don't think we've considered what would happen if a third party provides us with a "foreign" utxo that actually duplicates a utxo owned by the wallet, potentially causing us to sign something we didn't agree to. It might need more research so I'll open an issue. bitcoindevkit/bdk_wallet#29

I created a separated draft PR bitcoindevkit/bdk#1823 to address the issue.

nymius commented:
**review club notes 02/06/2025**

1. Include `psbt::Input::non_witness_utxo` field together with `psbt::Input::non_witness_utxo` when `txout.script_pubkey.witness_version().is_some()` while re constructing foreign UTxOs in `Wallet::build_fee_bump`.
3. The relevance of the integration of miniscript plan (#1786 ) for this PR was mentioned. `Wallet::get_available_utxos` obtained the satisfaction weight of UTxOs before passing them to `preselect_utxos`. As satisfaction weight is dependent of planning, there should be future work to integrate this, as it is a previous step to `Wallet::filter_utxos`.
4. The split of the PR was suggested, in two different ones to allow a faster review and merge of the fix for bitcoindevkit/bdk#1810.
5. As commits addressing suggested changes had been acknowledge by reviewers, the next step should be to squash them in four or less commits:
	1. Refactor
	2. `coin_selection::filter_duplicates` removal.
	3. Tests
	4. Miscellaneous not related to refactor
3. As part of the review we talked about bitcoindevkit/bdk_wallet#29  and its fix candidate bitcoindevkit/bdk#1823 :
	1. As bitcoindevkit/bdk#1823 introduces breaking changes, a compromise solution would be to add a silent check which works as a no-op in case transaction is found to be owned by the wallet. Also, one possible approach is to delay the introduction of the new `NotForeignUtxo` variant error (breaking change) until the release of the new major version, `2.0.0` with a `TODO` comment.
	2. It wasn't clear the ownership of the transactions stored in `TxGraph`. The terms `canonical` and `relevant` were mentioned as a good source to get an idea of the topic from the documentation.
	`Wallet::transactions` has a good explanation of this.
	`TxGraph`'s transactions don't have to belong to the wallet necessarily, and `get_txout` shouldn't be considered a good source of truth to establish if an `Outpoint` can be unlocked by `Wallet` or not. `Wallet::is_mine` is the correct method for this.

Thanks to @ValuedMammal, @oleonardolima, @LagginTimes and @evanlinjin for participating!

nymius commented:
**Next steps based on review club**:
- [x] Include `psbt::Input::non_witness_utxo` field together with `psbt::Input::non_witness_utxo` when `txout.script_pubkey.witness_version().is_some()` while re constructing foreign UTxOs in `Wallet::build_fee_bump`.
- [x] Squash commits.
- [x] Create separated PR to fix bitcoindevkit/bdk#1810

nymius commented:
905c8dd will be squashed to b1bff02fc3875bbb4342c3da694e26e9036bbb06 once bitcoindevkit/bdk#1830 is merged

nymius commented:
Now that foreign UTxOs are separated from manually selected UTxOs there are new cases for duplicity:
1. Duplicated UTxOs inside `foreign_utxos` field.
2. Duplicated UTxOs across `foreign_utxos` and required/optional UTxOs. Shouldn't be a problem once bitcoindevkit/bdk#1823 is merged. Because of the new check which avoids the addition of local UTxOs for `Wallet` as foreign UTxOs.

**TODO**:
- [x] Ensure `foreign_utxos` does not contain duplicated UTxOs: `TxParams::foreign_utxos` transformed into `HashSet` in  7cb0243 and added test to confirm property. Will be squashed to b1bff02 once change is ACKnowledge. The avoidance of duplicity across foreign and optional/required UTxOs will be guaranteed by changes in bitcoindevkit/bdk#1823

ValuedMammal commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -125,7 +125,8 @@ pub(crate) struct TxParams {
     pub(crate) external_policy_path: Option<BTreeMap<String, Vec<usize>>>,
-    pub(crate) utxos: Vec<WeightedUtxo>,
+    pub(crate) utxos: HashSet<LocalOutput>,
+    pub(crate) foreign_utxos: HashSet<WeightedUtxo>,
```
I'm worried that this won't be sufficient to check that the _outpoint_ is unique. For instance, one could change the satisfaction weight and get a different `WeightedUtxo` while using the same outpoint.

nymius replied:
Good point, will think more about this. Will reproduce your case in a test and use a `HashMap` with outpoint as key in the meantime.

nymius replied:
I modified the test slightly to cover that case. I didn't add a different one because it is not modifying the logic being tested, but just being more precise with the assertions.

I used `HashMap` instead of `BTreeMap` because we don't need the ordering features, however it can be argued that `BTreeMap` is already use it in `crates/wallet/src/wallet/tx_builder.rs` and we are adding a new import.
I prefered the former over the later.

nymius replied:
To avoid the duplication concern we had for #1582, we can implement `Hash` manually for `WeightedUTxO` and make it use the `Hash` implementation of `Outpoint`.

nymius replied:
I realized this won't be enough, as `HashSet` requires the following for keys `k1`, `k2`:
$k1 == k2 \Rightarrow \text{hash}(k1) == \text{hash}(k2)$.
But this should enforced by us. If we implement `Hash` as I said above and don't also modify `PartialEq` implementation to be compatible with this, the case where outpoint is the same but weight is different will introduce two different utxos in foreign utxos set.

I wouldn't modify `PartialEq` because is probably going to break other logic depending of the usual way of comparing two elements (i.e. considering all its fields).

`HashMap` is still the best solution so far.

nymius commented:
> [905c8dd](https://github.com/bitcoindevkit/bdk/commit/905c8dd663ee845911f52510fe40e58b22fed252) will be squashed to [b1bff02](https://github.com/bitcoindevkit/bdk/commit/b1bff02fc3875bbb4342c3da694e26e9036bbb06) once bitcoindevkit/bdk#1830 is merged

Squash done and rebased on master.

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1618,60 +1640,70 @@ impl Wallet {
+                                    utxo: Utxo::Foreign {
+                                        outpoint: txin.previous_output,
+                                        sequence: txin.sequence,
+                                        psbt_input: psbt_input_data,
```
I think this could also work, i.e. just set the `witness_utxo` based on the value of `witness_version`

```rust
    psbt_input: Box::new(psbt::Input {
        witness_utxo: txout
            .script_pubkey
            .witness_version()
            .map(|_| txout.clone()),
        non_witness_utxo: Some(prev_tx.as_ref().clone()),
        ..Default::default()
    }),
```

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2605,3 +2573,76 @@ macro_rules! doctest_wallet {
+            confirmation_time: 200,
+        };
+
+        insert_anchor(&mut wallet, txid, expected_anchor);
```
nit: It doesn't really make sense to anchor the txid to block 0. `insert_tx` should be enough to add the utxos to the wallet.

ValuedMammal commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -274,26 +275,14 @@ impl<'a, Cs> TxBuilder<'a, Cs> {
-                    wallet
-                        .get_utxo(*outpoint)
-                        .ok_or(AddUtxoError::UnknownUtxo(*outpoint))
+                    self.params.utxos.insert(outpoint);
```
nit: I would call this variable "output"
```suggestion
                    self.params.utxos.insert(output);
```

ValuedMammal commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -1098,4 +1090,61 @@ mod test {
+    fn not_duplicated_foreign_utxos_with_same_outpoint_but_different_weight() {
+        use crate::test_utils::get_funded_wallet_wpkh;
+        // Use two different wallets to avoid adding local utxos
+        let (wallet1, txid1) = get_funded_wallet_wpkh();
```
You may want to use wallets with different descriptors.
```suggestion
        let (wallet1, txid1) = get_funded_wallet_single(get_test_wpkh());
```

nymius commented:
Thanks @ValuedMammal! Updated and rebased

evanlinjin commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -274,26 +275,14 @@ impl<'a, Cs> TxBuilder<'a, Cs> {
-            }
-        }
+                .ok_or(AddUtxoError::UnknownUtxo(*outpoint))
+        });
```
Styling nit: Just use a for loop.

evanlinjin replied:
I think it's better to not insert anything when this errors.

evanlinjin commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -125,7 +125,8 @@ pub(crate) struct TxParams {
     pub(crate) external_policy_path: Option<BTreeMap<String, Vec<usize>>>,
-    pub(crate) utxos: Vec<WeightedUtxo>,
+    pub(crate) utxos: HashSet<LocalOutput>,
+    pub(crate) foreign_utxos: HashMap<OutPoint, WeightedUtxo>,
```
What would happen if we use a shared `HashMap`?

```rust
pub(crate) utxos: HashMap<OutPoint, WeightedUtxo>,
```

evanlinjin commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1976,117 +2003,52 @@ impl Wallet {
+                })
+                // only process UTxOs not selected manually, they will be considered later in the chain
+                // NOTE: this avoid UTxOs in both required and optional list
+                .filter(|may_spend| !params.utxos.contains(may_spend))
```
Why don't we filter out foreign utxos as well? It seems to make sense to put `utxos` and `foreign_utxos` in the same map?

nymius commented:
Addressed comments and rebased onto master.

evanlinjin commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -274,26 +274,28 @@ impl<'a, Cs> TxBuilder<'a, Cs> {
+                        )
+                    })
+            })
+            .collect::<Result<HashMap<OutPoint, WeightedUtxo>, AddUtxoError>>()
```
Nit: Can we collect this into a variable first? I think it'll make it easier to read.

evanlinjin commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2605,3 +2559,77 @@ macro_rules! doctest_wallet {
+                    },
+                )
+            })
+            .unwrap();
```
What is this for?

evanlinjin replied:
`params.utxos.insert` should be on the outside otherwise it makes it hard to read.

evanlinjin commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -274,26 +274,28 @@ impl<'a, Cs> TxBuilder<'a, Cs> {
     /// These have priority over the "unspendable" utxos, meaning that if a utxo is present both in
     /// the "utxos" and the "unspendable" list, it will be spent.
     pub fn add_utxos(&mut self, outpoints: &[OutPoint]) -> Result<&mut Self, AddUtxoError> {
```
Ideally, we would error here if an `OutPoint` conflicts with a foreign UTXO. However, we will need to add new error variants (which is a MAJOR change since the error type does not have `non-exhaustive`).

For now, how about we make a note in the docs that if you add the same UTXO as both foreign and local, the latest change has precedence (please word that better).

Same goes for the add-foreign-utxo methods.

evanlinjin replied:
For the add-foreign-utxo methods, ideally we will error out if it conflicts with pre-existing in `TxParams::utxos` or utxos in `TxBuilder::wallet`. However, the error variant is not `non-exhaustive` so that will be breaking.

For now, how about we just note that it would be unexpected behavior in the docs?

evanlinjin replied:
@nymius had a better idea of making local UTXOs have precedence over foreign UTXOs. Let's do this and also update the documentation.

nymius replied:
I've added the documentation requested and two tests to address the discussed cases. #1823 will fix the issue against tracked local UTxOs (the one the wallet already has knowledge about, aka `optional`).

evanlinjin reviewed (commented):
Documentation suggestions. Otherwise, LGTM

nymius commented:
Rebased onto b26ff891

nymius commented:
Rebased onto 362c3dc8

evanlinjin commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1618,60 +1626,71 @@ impl Wallet {
+                                    satisfaction_weight: self
+                                        .public_descriptor(keychain)
+                                        .max_weight_to_satisfy()
+                                        .unwrap(),
+                                    utxo: Utxo::Local(LocalOutput {
+                                        outpoint: txin.previous_output,
+                                        txout: txout.clone(),
+                                        keychain,
+                                        is_spent: true,
+                                        derivation_index,
+                                        chain_position,
+                                    }),
+                                },
+                            ),
+                            None => {
+                                let satisfaction_weight = Weight::from_wu_usize(
+                                    serialize(&txin.script_sig).len() * 4
+                                        + serialize(&txin.witness).len(),
+                                );
+
+                                (
+                                    txin.previous_output,
+                                    WeightedUtxo {
+                                        utxo: Utxo::Foreign {
+                                            outpoint: txin.previous_output,
+                                            sequence: txin.sequence,
+                                            psbt_input: Box::new(psbt::Input {
+                                                witness_utxo: txout
+                                                    .script_pubkey
+                                                    .witness_version()
+                                                    .map(|_| txout.clone()),
+                                                non_witness_utxo: Some(prev_tx.as_ref().clone()),
+                                                ..Default::default()
+                                            }),
+                                        },
+                                        satisfaction_weight,
+                                    },
+                                )
+                            }
                         }
```
**Nit:** trying to make this section fully functional has made it harder to read. There is weird `map` statements, and it becomes hard to see whether we are `map`-ing on an iterator or option.

I would prefer the following:

```rust
                let prev_tx = graph
                    .get_tx(txin.previous_output.txid)
                    .ok_or(BuildFeeBumpError::UnknownUtxo(txin.previous_output))?;
                let chain_position = chain_positions
                    .get(&txin.previous_output.txid)
                    .cloned()
                    .ok_or(BuildFeeBumpError::UnknownUtxo(txin.previous_output))?;
                let txout = prev_tx.output[txin.previous_output.vout as usize].clone();
                let (outpoint, weighted_utxo) = match txout_index
                    .index_of_spk(txout.script_pubkey.clone())
                {
                    Some(&(keychain, derivation_index)) => (
                        txin.previous_output,
                        WeightedUtxo {
                            satisfaction_weight: self
                                .public_descriptor(keychain)
                                .max_weight_to_satisfy()
                                .unwrap(),
                            utxo: Utxo::Local(LocalOutput {
                                outpoint: txin.previous_output,
                                txout: txout.clone(),
                                keychain,
                                is_spent: true,
                                derivation_index,
                                chain_position,
                            }),
                        },
                    ),
                    None => {
                        let satisfaction_weight = Weight::from_wu_usize(
                            serialize(&txin.script_sig).len() * 4 + serialize(&txin.witness).len(),
                        );
                        (
                            txin.previous_output,
                            WeightedUtxo {
                                utxo: Utxo::Foreign {
                                    outpoint: txin.previous_output,
                                    sequence: txin.sequence,
                                    psbt_input: Box::new(psbt::Input {
                                        witness_utxo: txout
                                            .script_pubkey
                                            .witness_version()
                                            .map(|_| txout.clone()),
                                        non_witness_utxo: Some(prev_tx.as_ref().clone()),
                                        ..Default::default()
                                    }),
                                },
                                satisfaction_weight,
                            },
                        )
                    }
                };
                Ok((outpoint, weighted_utxo))
```

nymius replied:
I agree, will try to control the bias towards functional style.

evanlinjin reviewed (approved):
ACK 2f83b4508fe2369c6336c9579517f3a4eb25d923

Just a nit for future reference.
